pkg/mutation/system.go

```diff
@@ -1,20 +1,173 @@
 package mutation
 
 import (
+	"fmt"
+	"sort"
 	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // System keeps the list of mutations and
 // provides an interface to apply mutations.
 type System struct {
-	// schemaDB schema.DB commented to please the linter, will uncomment in the implementation
-	mutators []Mutator
+	schemaDB        SchemaDB
+	orderedMutators []Mutator
+	mutatorsMap     map[ID]Mutator
 	sync.RWMutex
 }
 
 // NewSystem initializes an empty mutation system
 func NewSystem() *System {
 	return &System{
-		mutators: make([]Mutator, 0),
+		orderedMutators: make([]Mutator, 0),
+		mutatorsMap:     make(map[ID]Mutator),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	current, ok := s.mutatorsMap[m.ID()]
+	if ok && !m.HasDiff(current) {
+		return nil
+	}
+
+	toAdd := m.DeepCopy()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := toAdd.(MutatorWithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	s.mutatorsMap[toAdd.ID()] = toAdd
+
+	i := sort.Search(len(s.orderedMutators), func(i int) bool {
+		return greaterOrEqual(s.orderedMutators[i], toAdd)
+	})
+
+	if i == len(s.orderedMutators) { // Adding to the bottom of the list
+		s.orderedMutators = append(s.orderedMutators, toAdd)
+		return nil
+	}
+
+	found := equal(s.orderedMutators[i], toAdd)
+	if found {
+		s.orderedMutators[i] = toAdd
+		return nil
+	}
+
+	s.orderedMutators = append(s.orderedMutators, nil)
+	copy(s.orderedMutators[i+1:], s.orderedMutators[i:])
+	s.orderedMutators[i] = toAdd
+	return nil
+}
+
+// Mutate applies the mutation in place to the given object
+func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) error {
+	s.RLock()
+	defer s.RUnlock()
+
+	maxIterations := len(s.orderedMutators) + 1
+
+	old := obj.DeepCopy()
+
+	for i := 0; i < maxIterations; i++ {
+		for _, m := range s.orderedMutators {
+			if m.Matches(obj, ns) {
+				err := m.Mutate(obj)
+				if err != nil {
+					return errors.Wrapf(err, "Mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+				}
+			}
+		}
+		if cmp.Equal(old, obj) {
+			return nil
+		}
+		old = obj.DeepCopy()
+	}
+
+	return fmt.Errorf("Mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+}
+
+// Remove removes the mutator from the mutation system
+func (s *System) Remove(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if _, ok := s.mutatorsMap[m.ID()]; !ok {
```

Comment:
It doesn't look like we're removing the object from mutatorsMap anywhere in this function?

Reply:
Good catch, extending the test & fixing.